components/script/dom/toggleevent.rs

```diff
@@ -0,0 +1,143 @@
+        Self::new_with_proto(
+            window, None, type_, bubbles, cancelable, old_state, new_state, source, can_gc,
+        )
+    }
+
+    #[allow(clippy::too_many_arguments)]
+    fn new_with_proto(
+        window: &Window,
+        proto: Option<HandleObject>,
+        type_: Atom,
+        bubbles: EventBubbles,
+        cancelable: EventCancelable,
+        old_state: DOMString,
+        new_state: DOMString,
+        source: Option<DomRoot<Element>>,
+        can_gc: CanGc,
+    ) -> DomRoot<ToggleEvent> {
+        let event = Box::new(ToggleEvent::new_inherited(old_state, new_state, source));
+        let event = reflect_dom_object_with_proto(event, window, proto, can_gc);
+        {
+            let event = event.upcast::<Event>();
+            event.init_event(type_, bool::from(bubbles), bool::from(cancelable));
+        }
+        event
+    }
+}
+
+impl ToggleEventMethods<crate::DomTypeHolder> for ToggleEvent {
+    /// <https://html.spec.whatwg.org/multipage/#toggleevent>
+    fn Constructor(
+        window: &Window,
+        proto: Option<HandleObject>,
+        can_gc: CanGc,
+        type_: DOMString,
+        init: &ToggleEventBinding::ToggleEventInit,
+    ) -> Fallible<DomRoot<ToggleEvent>> {
+        let bubbles = EventBubbles::from(init.parent.bubbles);
+        let cancelable = EventCancelable::from(init.parent.cancelable);
+        Ok(ToggleEvent::new_with_proto(
+            window,
+            proto,
+            Atom::from(type_),
+            bubbles,
+            cancelable,
+            init.oldState.clone(),
+            init.newState.clone(),
+            init.source.as_ref().map(|s| DomRoot::from_ref(&**s)),
+            can_gc,
+        ))
+    }
+
+    /// <https://dom.spec.whatwg.org/#dom-event-istrusted>
+    fn IsTrusted(&self) -> bool {
+        self.event.IsTrusted()
+    }
+
+    /// <https://html.spec.whatwg.org/multipage/#dom-toggleevent-oldstate>
+    fn OldState(&self) -> DOMString {
+        self.old_state.clone()
+    }
+
+    /// <https://html.spec.whatwg.org/multipage/#dom-toggleevent-newstate>
+    fn NewState(&self) -> DOMString {
+        self.new_state.clone()
+    }
+
+    /// <https://html.spec.whatwg.org/multipage/#dom-toggleevent-source>
+    fn GetSource(&self) -> Option<DomRoot<Element>> {
+        // The source getter steps are to return the result of retargeting source against this's currentTarget.
+        let source = self.source.as_ref()?;
+
+        if let Some(current_target) = self.event.GetCurrentTarget() {
+            let retargeted = source.upcast::<EventTarget>().retarget(&current_target);
+            return retargeted.downcast::<Element>().map(DomRoot::from_ref);
+        }
+
+        let document = source.upcast::<Node>().GetOwnerDocument().unwrap();
+        let retargeted = source
+            .upcast::<EventTarget>()
+            .retarget(document.upcast::<EventTarget>());
+        retargeted.downcast::<Element>().map(DomRoot::from_ref)
```

Comment:
Can you file a spec bug that this should be updated in the spec given all major browsers implement it this way?

Reply:
Yeah I'll double check what the state of things is and file an appropriate issue.